File: versus/js/05-asynchronous-computation/os-events.cc
#include <boost/asio.hpp>
#include <csignal>
#include <iostream>

#define BOOST_THREAD_VERSION 4
#include <boost/thread/future.hpp>

using namespace std;
namespace asio = ::boost::asio;

int main() {
	asio::io_context ioc;

	asio::signal_set terminationSignals(ioc, SIGINT, SIGTERM);

	auto shutdown = [&](const boost::system::error_code& ec, int ) {
		if (!ec) {
			terminationSignals.cancel();
		}
	};

	terminationSignals.async_wait(shutdown);

	asio::post(ioc, [](){
		cout << "Hello World!" << endl;
	});

	cout << "setup done." << endl;

	// This method blocks the main thread.
	ioc.run();

	cout << "exiting..." << endl;

	return 0;
}

File: versus/js/05-asynchronous-computation/os-events.js
/****** Browser ******/

// Nothing to do here: The process runs as long as the
// browser tab is not closed.

setTimeout(function(){
    console.log("Hello Browser-World!");
}, 0);

/******Node.js *******/

// run for 11 days
const runTimer = setTimeout(function keepAlive(){}, 1.0e+9);

function shutdown() {
	console.log("exiting...");
	clearTimeout(runTimer);
}

process.on('SIGINT', shutdown);
process.on('SIGTERM', shutdown);

process.nextTick(function(){
    console.log("Hello Node.js-World!");
});

console.log("setup done.");
